rfcs/20240808-graph-api-int-compression-for-sdpa/README.md

```diff
@@ -0,0 +1,307 @@
+logical_tensor scale_desc {
+        1, data_type::f32, scale_dims, layout_type::strided};
+logical_tensor dst_desc {2, data_type::f32, src_dims, layout_type::strided};
+
+op deq = op(0, op::kind::DynamicDequantize, {src_desc, scale_desc},
+        {dst_desc}, "deq");
+// Specify the quantization type as per_group quantization.
+deq.set_attr<std::string>(op::attr::qtype, "per_group");
+// Group size indicates the the size of each group.
+deq.set_attr<int64_t>(op::attr::group_size, G);
+
+// Create graph and add the op to the graph
+graph g(dnnl::engine::kind::gpu);
+g.add_op(deq);
+
+// ...
+```
+
+
+#### Alternative options
+
+Similar to the discussion of int8 and fp8 data types, an alternative solution
+may be supporting int4 data types directly in computation operations like MatMul
+and Convolution. But it will bloat and complicate the opset and op schema of
+oneDNN Graph. Hence it's not considered here.
+
+In addition, currrently oneDNN Graph does not support the direct int8
+quantization and dequantization between int8 and f16/bf16 data types. To align
+with current integration flow and maintain the consistency of library design,
+quantization and dequantization between int4 and f16/bf16 data types will also
+be implemented with a combination of quantization ops and `TypeCast`. An
+alternative solution may be to extend quantization ops to support data type
+conversion directly between int4 and f16/bf16 data types. But it will
+unnecessarily complicates the fusion patterns in oneDNN Graph. For example,
+an int4 quantization will be expressed in two different ways: 
+
+1. With a single `Quantize(f16 -> int4)`
+2. With two operations `TypeCast(f16 -> f32) -> Quantize(f32 -> int4)`
+
+It will require extensions for all patterns related to mixed data type
+quantization. Hence the option is not considered here.
+
+### Proposal 3: Extend SDPA pattern with compressed K/V
+
+#### Pattern definition
+
+With the reference to int4 GQA pattern in
+[PyTorch](https://pytorch.org/blog/int4-decoding/), the proposed fusion pattern
+for SDPA with compressed K/V in oneDNN Graph API is defined as:
+
+![alt text](img/sdpa_pattern.png)
+
+As illustrated in the figure above, K cache and V cache are stored on user side
+in int4 or int8 data types. `DynamicDequantize` is used to convert int4 data to
+f32, then an optional `TypeCast` is used to convert f32 data types to f16 or
+bf16 data types when needed. Two `MatMul`s are used to compute the matrix
+multiplication of Q & K, and score & V. oneDNN Graph will fuse the pattern into
+a single partition with optimized CPU and GPU implementations.
+
+To get rid of the ambiguity of the computation type of the MatMul ops in the
+pattern and give users the control of leveraging higher precision data types to
+keep the model accuracy, we propose to follow the design of floating point math
+mode in primitive API. Once the math mode is specified, the backend will
```

Comment:
There's a Boolean flag to force integer up-conversion paired with a targeted data type to which it should be up-converted to. How is it going to be implemented for a Graph API? Could you add a code sample, please?

Reply:
OK. I have extended the Graph API with an independent setter function to accommodate  the boolean flag.